securesystemslib/storage.py

```diff
@@ -233,14 +240,34 @@ def __exit__(self, exc_type, exc_val, traceback):
   get = GetFile
 
 
-  def put(self, fileobj, filepath):
+  def put(self, fileobj, filepath, mode=None):
     # If we are passed an open file, seek to the beginning such that we are
     # copying the entire contents
     if not fileobj.closed:
       fileobj.seek(0)
 
+    # When calculating file permissions, the current OS umask value is first
+    # masked out. If custom permissions are set, we calculate a new umask
+    # followed by a call to open() with default permissions (0o777).
+    # This trick is done since open() does not support mode(permissions) as
+    # input parameter, alternatively os.open() can be used.
+
+
+    # If a file with the same name already exists, the new permissions
+    # may not be applied.
     try:
-      with open(filepath, 'wb') as destination_file:
+      os.remove(filepath)
+    except OSError:
+      pass
+
+
+    try:
+      if mode is None:
+        fd = os.open(filepath, os.O_WRONLY|os.O_CREAT)
+      else:
+        fd = os.open(filepath, os.O_WRONLY|os.O_CREAT, mode)
+
```

Comment:
I don't like these lines, it looks ugly.
My other option was to use `0o777 `as a default value for mode same as` os.open() `but I wanted to avoid setting a concrete value and leave it to the `os.open` implementation.
Let me know if you have different ideas.

Reply:
seems ok to me.

Question on fd though: does fd get closed by `with os.fdopen()`? It seems like that would make sense but please check

Reply:
The file descriptor seems to be closed together with the file object, at least my tests confirm it. Couldn't find it documented really.

Reply:
It makes sense that it would, I was just suspicious :)